### ISSUE TITLE
msp430 build fails with "LLVM ERROR: Cannot Select..."

### ISSUE THREAD

**Issue by nmichaels** [ziglang/zig#7346]
After applying the following patch:
```
--- a/lib/std/special/compiler_rt/atomics.zig   Tue Dec 08 00:07:14 2020 -0700
+++ b/lib/std/special/compiler_rt/atomics.zig   Tue Dec 08 14:21:13 2020 -0500
@@ -53,19 +53,19 @@
 // Generic version of GCC atomic builtin functions.
 // Those work on any object no matter the pointer alignment nor its size.
 
-fn __atomic_load(size: u32, src: [*]u8, dest: [*]u8, model: i32) callconv(.C) void {
+fn __atomic_load(size: c_uint, src: [*]u8, dest: [*]u8, model: i32) callconv(.C) void {
     var sl = spinlocks.get(@ptrToInt(src));
     defer sl.release();
     @memcpy(dest, src, size);
 }
 
-fn __atomic_store(size: u32, dest: [*]u8, src: [*]u8, model: i32) callconv(.C) void {
+fn __atomic_store(size: c_uint, dest: [*]u8, src: [*]u8, model: i32) callconv(.C) void {
     var sl = spinlocks.get(@ptrToInt(dest));
     defer sl.release();
     @memcpy(dest, src, size);
 }
 
-fn __atomic_exchange(size: u32, ptr: [*]u8, val: [*]u8, old: [*]u8, model: i32) callconv(.C) void {
+fn __atomic_exchange(size: c_uint, ptr: [*]u8, val: [*]u8, old: [*]u8, model: i32) callconv(.C) void {
     var sl = spinlocks.get(@ptrToInt(ptr));
     defer sl.release();
     @memcpy(old, ptr, size);
@@ -73,7 +73,7 @@
 }
 
 fn __atomic_compare_exchange(
-    size: u32,
+    size: c_uint,
     ptr: [*]u8,
     expected: [*]u8,
     desired: [*]u8,
```

and stuffing the following code into a file called blink.zig:

```zig
const std = @import("std");

extern var __P1DIR: u8;
const wdtctl = @intToPtr(*volatile u16, 0x120);
const p1out = @intToPtr(*volatile u8, 0x0021);
const p1dir = @intToPtr(*volatile u8, 0x0022);

const wdtpw: u16 = 0x5a00;
const wdthold: u16 = 0x0080;

const loop_ms: u16 = 0x100;

fn delay(ms: u8) void {
    var count: u8 = 0;
    while (count < ms) : (count += 1) {
        var loop: u16 = 0;
        while (loop < loop_ms) : (loop += 1) {
            asm volatile (""
                :
                :
                : "memory"
            );
        }
    }
}

export fn main() noreturn {
    wdtctl.* = wdtpw + wdthold;
    p1dir.* |= 0x01;
    p1out.* |= 0x01; // Turn on the LED.
    while (true) {
        delay(100);
        p1out.* ^= 1;
    }
}
```

I get the following build error:

```
$ zig build-exe --strip -O ReleaseSmall -target msp430-freestanding-none -mcpu=msp430 blink.zig
LLVM ERROR: Cannot select: 0x55b4c9266770: ch = AtomicStore<(store release 2 into %ir.6, align 64)> 0x55b4c9309f28:1, 0x55b4c9ee06a0, Constant:i16<0>
  0x55b4c9ee06a0: i16,ch,glue = CopyFromReg 0x55b4c9d565c0, Register:i16 $r12, 0x55b4c9d565c0:1
    0x55b4c9edef70: i16 = Register $r12
    0x55b4c9d565c0: ch,glue = callseq_end 0x55b4c97a9508, TargetConstant:i16<0>, TargetConstant:i16<0>, 0x55b4c97a9508:1
      0x55b4c8fa2f80: i16 = TargetConstant<0>
      0x55b4c8fa2f80: i16 = TargetConstant<0>
      0x55b4c97a9508: ch,glue = MSP430ISD::CALL 0x55b4c9c7b010, TargetGlobalAddress:i16<%compiler_rt.atomics.Spinlock* (i16)* @compiler_rt.atomics.SpinlockTable.get> 0, Register:i16 $r12, 0x55b4c9c7b010:1
        0x55b4c930d6a0: i16 = TargetGlobalAddress<%compiler_rt.atomics.Spinlock* (i16)* @compiler_rt.atomics.SpinlockTable.get> 0
        0x55b4c9edef70: i16 = Register $r12
        0x55b4c9c7b010: ch,glue = CopyToReg 0x55b4ca4137e0, Register:i16 $r12, 0x55b4c97250a0
          0x55b4c9edef70: i16 = Register $r12
          0x55b4c97250a0: i16,ch = CopyFromReg 0x55b4c82746f8, Register:i16 %1
            0x55b4c9387e40: i16 = Register %1
  0x55b4c9b67450: i16 = Constant<0>
In function: __atomic_load
Aborted
```

**Comment by LemonBoy** [comment]
I solved a similar problem in #4959.

The idea is that, if the platform doesn't support any kind of atomics (eg. MSP430 or ARM's Cortex-M0) we can gracefully fall back to something suboptimal but that, if invoked, works.
By setting `largest_atomic_cas_size` to `0` for ARM `v6m` targets we force the use of locking via artisanal spinlocks that, to do their job, require `@atomicRmw` to work.
The ARM backend produces a call to `__sync_lock_test_and_set_4` that can be easily implemented by the user, the MSP backend is emitting the error you're seeing.

The solutions are:
* Revert the `largest_atomic_cas_size=0` trick and simply avoid exporting any atomic-related builtin for the aforementioned platforms. Using an atomic operation becomes a link-time error but, on the plus side, we don't add extra complexity;
* Tell LLVM devs about this bug and wait for their decision on the matter, if they decide against emitting calls to `__sync` builtins we're back to square one

**Comment by zigazeljko** [comment]
For such platforms, you don't actually need spinlocks, you can simply disable interrupts before doing the operation and enable them afterwards (if they were enabled before).

In fact, this is the only correct way, since using spinlocks inside an interrupt handler would deadlock anyway.
